Verifies able to select a single resultset

rbillings opened the pull request:
Fixes issue #4

rbillings commented:
@mozilla/web-qa-sorcerers r?

davehunt commented on pages/resultset.py:
```diff
@@ -12,6 +12,10 @@
+    _page_title = '[0] mozilla-inbound'
+
+    _first_resultset_locator = (By.CSS_SELECTOR, '.th-view-content > ng-view:nth-child(1) > div:nth-child(2) > div:nth-child(1) > span:nth-child(1) > span:nth-child(1) > span:nth-child(2) > a:nth-child(1)')
+    _first_resultset_number_locator = (By.CSS_SELECTOR, '.th-view-content > ng-view:nth-child(1) > div:nth-child(2) > div:nth-child(3) > div:nth-child(2) > span:nth-child(1) > ul:nth-child(1) > li:nth-child(2) > span:nth-child(1) > span:nth-child(1) > a:nth-child(1)')
```
This isn't really a number, but a revision. Also, this is a rather complex locator. Looking at the DOM, I think this would be suitable: `.result-set .revision-holder a`.

Also, this locator isn't specific to the first match, so we don't need to prefix it with `first_`, we just only use the first match in the test/page object.

davehunt commented on pages/resultset.py:
```diff
@@ -12,6 +12,10 @@
 
+    _page_title = '[0] mozilla-inbound'
+
+    _first_resultset_locator = (By.CSS_SELECTOR, '.th-view-content > ng-view:nth-child(1) > div:nth-child(2) > div:nth-child(1) > span:nth-child(1) > span:nth-child(1) > span:nth-child(2) > a:nth-child(1)')
```
I think we should name this `_resultset_link_locator` and the following would be a less fragile locator: `.result-set .result-set-title-left > span a`.

davehunt commented on pages/resultset.py:
```diff
@@ -32,6 +36,15 @@ def job_result_status(self):
     def unclassified_failure_count(self):
         return int(self.selenium.find_element(*self._unclassified_failure_count_locator).text)
 
+    def get_first_resultset_number(self):
```
This should be a property, drop the `get` and change `number` to `revision`.

davehunt commented on tests/test_view_single_result.py:
```diff
@@ -0,0 +1,22 @@
+        resultsetnumber = resultset_page.get_first_resultset_number()
+        resultset_page.open_single_resultset()
+
+        assert resultset_page.get_url_current_page().endswith(resultsetnumber)
```
I don't think we need to abstract the Selenium property for the current URL. This can simply be `assert selenium.current_url.endswith(revision)`

davehunt commented:
Sorry for the delay in review this @rbillings. You might want to consider creating a page region for the result set, as it's a repeating pattern. That way you could have a property for getting the revision and a method for viewing the result in isolation, and could reference the first result set via an index. Here's what the test might look like with the suggested changes.

```python
@pytest.mark.nondestructive
def test_open_single_result(self, base_url, selenium):
    resultset_page = ResultsetPage(base_url, selenium).open()
    revision = resultset_page.resultset[0].revision
    resultset_page.revision[0].view()
    assert selenium.current_url.endswith(revision)
```

rbillings commented:
Thanks for the suggestions @davehunt. I updated the names and locators. I also changed the test verification as I realized that some revisions have multiple elements, and the only unique factor for each revision is the date/time - which also better matches the original test case.

davehunt commented:
It looks like this needs a rebase, and also Travis seems to have had an issue.

davehunt commented on pages/resultset.py:
```diff
@@ -12,10 +12,12 @@
     _job_result_status_locator = (By.CSS_SELECTOR, '#result-status-pane > div:nth-child(1) > span')
     _logviewer_button_locator = (By.ID, 'logviewer-btn')
-    _resultset_locator = (By.CSS_SELECTOR, 'div.row.result-set')
+    _resultset_locator = (By.CSS_SELECTOR, '.result-set .result-set-title-left > span a')
```
It's not clear from the name that this is a locator for the date/timestamp now.

davehunt commented:
I still think a ResultSet page region makes sense here. I've updated my example to take into account the switch to the date/timestamp:

```python
@pytest.mark.nondestructive
def test_open_single_result(self, base_url, selenium):
    results_page = ResultsetPage(base_url, selenium).open()
    timestamp = results_page.results[0].timestamp
    single_result_page = results_page.results[0].view()
    assert 1 == len(single_result_page.results)
    assert timestamp == single_result_page.results[0]
```

rbillings commented:
@davehunt I absolutely agree that there should be a ResultSet page region. I did not make any changes in this PR for that as the other open PR contains many page region additions that would be needed here. After that PR is merged I will add the updates for creating a ResultSet page region. Thanks for the review!

rbillings commented:
The last update has a rebase, added page region and other fixes suggested. I did not add the suggested timestamp as the files have timestamps which are important- more than what time it is when the tests are run.

davehunt commented on pages/page.py:
```diff
@@ -1,3 +1,5 @@
+#!/usr/bin/env python
+
```
We don't need shebangs in files that are not going to be executed directly on the command line.

davehunt commented on pages/page.py:
```diff
@@ -19,6 +21,9 @@ def __init__(self, base_url, selenium, timeout=10, **kwargs):
 
+    def go_to_page(self):
+        self.open('')
+
```
This is unused, and should be removed.

davehunt commented on pages/treeherder.py:
```diff
@@ -13,11 +13,11 @@
 
 class TreeherderPage(Base):
 
+    _first_resultset_locator = (By.CSS_SELECTOR, '.th-view-content > ng-view:nth-child(1) > div:nth-child(2) > div:nth-child(1) > span:nth-child(1) > span:nth-child(1) > span:nth-child(2) > a:nth-child(1)')
```
This looks like a very fragile locator, and also doesn't appear to be used.

davehunt commented on pages/treeherder.py:
```diff
@@ -64,3 +74,22 @@ def __init__(self, base_url, selenium):
+    def __init__(self, base_url, selenium):
+        Page.__init__(self, base_url, selenium)
+        Wait(self.selenium, self.timeout).until(
+            expected.window_with_title('mozilla-inbound'))
```
This has a hard-coded branch name, which I suspect is not ideal.

davehunt commented:
> I did not add the suggested timestamp as the files have timestamps which are important- more than what time it is when the tests are run.

Sorry, I wasn't able to understand this, but perhaps it doesn't matter. The branch needs updating from master, and I've added a few comments.

rbillings commented:
closing/re-opening to trigger travis

rbillings commented:
@mozilla/web-qa-sorcerers r?

stephendonner commented:
I don't think my looking at this qualifies as a review, but I will comment that I got it to run locally without issue, and created our first ad-hoc job for Treeherder, here, and ran this pull through that, without issue (once I figured out the config): http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/treeherder.adhoc/9/console

But again, I'm just getting back into the automation "game," so I've got very little skin :-(

davehunt commented on pages/treeherder.py:
```diff
@@ -64,3 +73,22 @@ def __init__(self, base_url, selenium):
         return self.is_element_visible(self._job_header_locator)
+
+
+class SingleResultPage(Page):
```
Perhaps with the exception of the `__init__` this page object looks like it could be rolled up into `TreeherderPage`. The main results page shows > 1 result, and the single result page appears to be exactly the same in terms of structure, but only has one result. Therefore a property like `results_count` can apply to both. For the revision date, the existing `first_revision_date` would also work. Perhaps this can be addressed later, but I think we should at least improve the `_single_resultset_datestamp_locator` here, as it appears more fragile than `_first_resultset_datestamp_locator` even though they locate the same element.

rbillings replied:
It's true that the format of the page does not change, but the url does- which is why a page region made sense to me. You cannot view the single result without an action. You could also say the same about the logviewer - it is part of the treeherder page but it only displays with an action. It's confusing to me.

Lastly I totally agree about the locators.

davehunt replied:
Only the query string changes in the URL, the two pages most likely share the same view in the application. The log viewer has a very different DOM, and so I wouldn't consider it comparable to the result sets page.

You could have `SingleResultPage` extending `TreeherderPage` and then you'd only need to implement locators and functions that are specific to when there's a single result show in the page. That said, I suspect anything that's relevant to a single result shown is also relevant to multiple results.

davehunt commented on tests/test_view_single_result.py:
```diff
@@ -0,0 +1,19 @@
+
+    @pytest.mark.nondestructive
+    def test_open_single_result(self, base_url, selenium):
+
```
Nit: Remove this blank line.

rbillings commented:
I've addressed all previous comments. Ready for review.

davehunt commented on pages/treeherder.py:
```diff
@@ -13,11 +13,12 @@
-    _resultset_locator = (By.CSS_SELECTOR, 'div.row.result-set')
-    _result_status_locator = (By.ID, 'job-details-panel')
+    _results_locator = (By.CSS_SELECTOR, '.result-set-bar')
+    _single_resultset_datestamp_locator = (By.CSS_SELECTOR, '.result-set-title-left > span:nth-child(2) > a:nth-child(1)')
```
When there's only one resultset, our `_first_resultset_datestamp_locator` should still work.

davehunt commented on pages/treeherder.py:
```diff
@@ -33,8 +34,20 @@ def job_result_status(self):
+
+    @property
+    def revision_date(self):
+        return self.selenium.find_element(*self._single_resultset_datestamp_locator).text
```
I think we can just call `first_revision_date`, it just so happens that there's only one result, so the first result is still accurate.

davehunt commented on pages/treeherder.py:
```diff
@@ -51,6 +64,12 @@ def open_perfherder_page(self):
+        Wait(self.selenium, self.timeout).until(
+            EC.visibility_of_element_located(self._first_resultset_datestamp_locator))
+        self.selenium.find_element(*self._first_resultset_datestamp_locator).click()
+        return self.selenium.find_element(*self._single_resultset_datestamp_locator)
```
You don't need to return anything here. The only thing that might be worth returning is a page object for the single result page, for which you could just return `self` as it's the same as the existing page (but with less results).

davehunt commented on tests/test_view_single_result.py:
```diff
@@ -0,0 +1,18 @@
+from pages.treeherder import TreeherderPage
+
+
+class TestSingleResult:
```
I'm going to raise this as a separate issue (and mark it as suitable for Outreachy applicants), but we really don't need test classes in pytest. In unittest they're required because you inherit from a base `TestClass` that gives you helpers for assertions, as well as other things. In pytest you can just define test functions.

davehunt commented:
Looking great @rbillings! Just a couple of comments, with those addressed this has my r+

rbillings commented:
@stephendonner - after checks are run, if you could review that would be great

rbillings commented:
@mozilla/web-qa-sorcerers r?

stephendonner commented:
Thx @rbillings - has addressed Dave's comments, runs for me locally, and also in the ad-hoc run: http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/treeherder.adhoc/10/console

r+

stephendonner commented:
I think we can merge this after squashing these 8 commits, right?

davehunt commented:
Rebased, squashed, and merged: https://github.com/mozilla/treeherder-tests/commit/ad19cac05d0c29e34e2edaee0e306558c3b3c750
